osu.Server.Spectator/Hubs/MultiplayerHubContext.cs

```diff
@@ -128,26 +128,27 @@ public async Task ChangeAndBroadcastUserState(ServerMultiplayerRoom room, Multip
         {
             log(room, user, $"User state changed from {user.State} to {state}");
 
-            user.State = state;
-
             string? connectionId = users.GetConnectionIdForUser(user.UserID);
 
             if (connectionId != null)
             {
-                switch (state)
+                if (state == MultiplayerUserState.FinishedPlay)
                 {
-                    case MultiplayerUserState.FinishedPlay:
-                    case MultiplayerUserState.Idle:
-                        await context.Groups.RemoveFromGroupAsync(connectionId, MultiplayerHub.GetGroupId(room.RoomID, true));
-                        break;
-
-                    case MultiplayerUserState.Ready:
-                    case MultiplayerUserState.Spectating:
+                    // when entering FinishedPlay state, we can be confident that player can be removed from group
+                    await context.Groups.RemoveFromGroupAsync(connectionId, MultiplayerHub.GetGroupId(room.RoomID, true));
+                }
+                else
+                {
+                    // when leaving Results state, we can be confident that player can be removed from group
+                    if (user.State == MultiplayerUserState.Results)
+                    {
                         await context.Groups.AddToGroupAsync(connectionId, MultiplayerHub.GetGroupId(room.RoomID, true));
-                        break;
+                    }
                 }
             }
 
+            user.State = state;
```

Comment:
If you *need* to use the previous state, make a variable called `previousState`. Then put this back where it was.

Reply:
using the previous state was the only way I could find to keep `EnsureThereIsNoGameplayGroupMemoryLeak` test passing. `Spectating` state throws everything off.